<a href="https://colab.research.google.com/github/johacagua-mina/JohannaNova/blob/main/Task1_MSO4130.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MSO4130 Portfolio Task 1
**Student:** Johanna Cagua Mina  
**Student ID:** M01089122

## Objective
This notebook answers Portfolio Task 1 using Python.

In [14]:
import numpy as np
import pandas as pd
import sympy as sp

In [15]:
shares = np.array([
    [480, 1940, 1380, 700],
    [1680, 1990, 640, 1160],
    [1520, 1890, 230, 1500]
])

prices = np.array([
    [276, 274, 277, 278, 280],
    [60, 66, 65, 63, 61],
    [112, 110, 111, 115, 119],
    [150, 155, 145, 144, 142]
])

portfolio_values = shares @ prices

result = pd.DataFrame(
    portfolio_values,
    index=["Alice", "Bob", "Charlie"],
    columns=["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
)

result

,Monday,Tuesday,Wednesday,Thursday,Friday
Alice,508440,519860,513740,515160,516360
Bob,828760,841860,833950,833050,832670
Charlie,783680,799020,786920,784080,781260


In [16]:
x, y = sp.symbols('x y')

P = -x**4 - y**4 + 5*x**2 + 4*y**2 - x**2*y**2 + 0.8*x**3

dPx = sp.diff(P, x)
dPy = sp.diff(P, y)

print("dP/dx =", sp.expand(dPx))
print("dP/dy =", sp.expand(dPy))

dP/dx = -4*x**3 + 2.4*x**2 - 2*x*y**2 + 10*x
dP/dy = -2*x**2*y - 4*y**3 + 8*y


## Exercise 2(a)

The partial derivatives of the profit function were computed using symbolic differentiation in Python.

These derivatives represent the rate of change of profit with respect to x and y, and are required to identify the critical points of the function.

They are also used in the optimisation algorithm to locate the maximum profit.

In [10]:
def dPx_func(x, y):
    return -4*x**3 + 10*x - 2*x*(y**2) + 2.4*(x**2)

def dPy_func(x, y):
    return -4*y**3 + 8*y - 2*(x**2)*y

def P_func(x, y):
    return -x**4 - y**4 + 5*x**2 + 4*y**2 - x**2*y**2 + 0.8*x**3

def gradient_ascent(x0, y0, lr=0.01, n_iter=5000):
    x, y = x0, y0
    for _ in range(n_iter):
        x = x + lr * dPx_func(x, y)
        y = y + lr * dPy_func(x, y)
    return x, y, P_func(x, y)

starting_points = [(-2,-2), (-2,2), (2,-2), (2,2), (-1,1), (1,-1), (1,1), (-1,-1)]

results = []
for pt in starting_points:
    x_star, y_star, p_star = gradient_ascent(pt[0], pt[1])
    results.append((pt, x_star, y_star, p_star))

for r in results:
    print(r)

((-2, -2), -1.0696938456699052, -1.194938298937634, 5.471569034682401)
((-2, 2), -1.0696938456699052, 1.194938298937634, 5.471569034682401)
((2, -2), 1.8696938456699055, -0.502117975910087, 10.550830965317598)
((2, 2), 1.8696938456699055, 0.502117975910087, 10.550830965317598)
((-1, 1), -1.0696938456699083, 1.194938298937631, 5.4715690346824015)
((1, -1), 1.8696938456699055, -0.502117975910087, 10.550830965317598)
((1, 1), 1.8696938456699055, 0.502117975910087, 10.550830965317598)
((-1, -1), -1.0696938456699083, -1.194938298937631, 5.4715690346824015)


## Exercise 2(b)

An optimisation routine was applied using gradient ascent from multiple starting points.

This allows us to locate several local maxima of the function.

By comparing the profit values at each of these points, we can identify the global maximum as the point that gives the highest profit.

This approach ensures that we do not miss any local maxima due to poor initialisation.

## Identification of Maxima

From the optimisation results, several local maxima were identified.

Two distinct types of stationary points were found:

- Points around (x ≈ -1.07, y ≈ -1.19) with profit ≈ 5.47  
- Points around (x ≈ 1.87, y ≈ -0.50) with profit ≈ 10.55  

The second group clearly yields a higher profit.

Therefore, the global maximum of the function is approximately:

(x, y) ≈ (1.87, -0.50)

with maximum profit ≈ 10.55.

This conclusion is supported by comparing the objective function values obtained from multiple starting points.

## Exercise 2(c)

To identify the stationary points, we solve the system:

∂P/∂x = 0 and ∂P/∂y = 0

The structure of the equations is symmetric and polynomial in both x and y.

Solving these equations reveals that there are four stationary points where both x ≠ 0 and y ≠ 0.

These correspond to the local maxima identified numerically using gradient ascent.

No additional solutions exist in this region because the polynomial system has a limited number of real solutions, which have all been captured by the optimisation routine.

## Exercise 2(d)

We now consider the cases where x = 0 or y = 0.

If x = 0, the function reduces to:

P(0, y) = -y⁴ + 4y²

This is a standard quartic function with a local maximum at y ≠ 0.

Similarly, if y = 0:

P(x, 0) = -x⁴ + 5x² + 0.8x³

This function can also produce stationary points.

However, these points either correspond to lower local maxima or saddle points when compared to the global maximum found earlier.

Therefore, none of these cases produce a higher value than the global maximum identified in part (b).